codalab/server/bundle_manager.py

```diff
@@ -333,36 +336,42 @@ def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run, user_
+                if bundle.metadata.request_queue:
+                    workers_list = self._get_matched_workers(
+                        bundle.metadata.request_queue, user_owned_workers
+                    )
+                else:
+                    workers_list = user_owned_workers
                 return workers_list
 
-            workers_list = None
+            workers_list = []
             if bundle.owner_id != self._model.root_user_id:
-                # First try private workers
-                workers_list = get_available_workers(bundle.owner_id)
-                # If there is no user_owned worker, try to schedule the current bundle to run on a CodaLab's public worker.
-                if len(workers_list) == 0:
-                    # Check if there is enough parallel run quota left for this user
-                    if (
-                        self._model.get_user_parallel_run_quota_left(
-                            bundle.owner_id, user_info_cache[bundle.owner_id]
-                        )
-                        <= 0
-                    ):
-                        logger.info(
-                            "User %s has no parallel run quota left, skipping job for now",
-                            bundle.owner_id,
-                        )
-                        # Don't start this bundle yet, as there is no parallel_run_quota left for this user.
-                        continue
-            if not workers_list:
-                # Length is 0 (private user with no workers) or is None (root user)
-                workers_list = get_available_workers(self._model.root_user_id)
+                # First try workers that belong to the owner of the current bundle
+                workers_list = filter_workers_by_request_queue(bundle.owner_id)
+
+            # If there is no user owned worker or the bundle is requested to run on a CodaLab-owned worker,
+            # try to schedule the current bundle to run on a CodaLab-owned worker.
+            if len(workers_list) == 0:
+                # Check if there is enough parallel run quota left for this user
+                if (
+                    self._model.get_user_parallel_run_quota_left(
+                        bundle.owner_id, user_info_cache[bundle.owner_id]
+                    )
+                    <= 0
+                ):
+                    logger.info(
+                        "User %s has no parallel run quota left, skipping job for now",
+                        bundle.owner_id,
+                    )
+                    # Don't start this bundle yet, as there is no parallel_run_quota left for this user.
+                    continue
+                # Try to matched workers in CodaLab-owned instances
+                workers_list = filter_workers_by_request_queue(self._model.root_user_id)
 
+            workers_list = self._deduct_worker_resources(workers_list, running_bundles_info)
```

Comment:
If there are private workers returned but they don't match the resource requirements, it seems like we don't have a chance to add back the codalab-owned workers?  In general, I find this logic difficult to follow and it would be nice if we simply 1) got all the workers (public and private), 2) filtered out the ones which don't satisfy the resource constraints, and 3) prioritized all the workers.

Reply:
You can still do the parallel quota check after getting all the relevant workers first (since for each worker, you know whether it's public or not), which would happen in step 2).  You might lose a bit of performance by doing the 1) 2) 3) thing I suggested, but I think it's worth making the code easier to read so it's easy to reason about first, and then we can consider optimizations.  @bkgoksel can weigh in here too.

Reply:
Sure. Sounds good. Will work on this~